### DISCUSSION TITLE
Measure request network latency

### DISCUSSION THREAD

**Discussion by nflaig** [fastify/fastify#5226] [Q&A]
I was looking at the available hooks that fastify provides but it is not clear to me what hook should be used to measure the request network latency, i.e. when request is received and the request body is fully streamed already.

As per my current understanding the onRequest (or alternatively the preParsing) hook should be the ones to use for this. As based on [preParsing docs](https://fastify.dev/docs/latest/Reference/Hooks/#preparsing) you can modify the request payload stream in that hook which means it should be already streamed, or at least that's what I would expect.

For example looking at these logs
```
07:38:50.098            debug: onRequest req-4a1 127.0.0.1 publishBlockV2
07:38:50.098            debug: preParsing req-4a1 127.0.0.1 publishBlockV2
07:38:53.669            debug: preValidation req-4a1 127.0.0.1 publishBlockV2
07:38:53.669            debug: preHandler req-4a1 127.0.0.1 publishBlockV2
07:38:53.693            debug: Res req-4a1 publishBlockV2 - 200
```

The time between onRequest and preValidation (about 3.5 seconds) is only related to **Parsing** ([request lifecycle](https://fastify.dev/docs/latest/Reference/Lifecycle/#lifecycle)) and not network latency or the body being streamed.

**Comment by jsumners** [comment]
If you're trying to determine the time it takes to process a request before it is given over to the handler: `onRequest` and then `preHandler`.

**Reply by nflaig** [reply]
This is clear to me but at which point is the request fully received (excluding any parsing / validation). Is it `onRequest` then, is the body fully streamed at this point in time?

**Reply by jsumners** [reply]
It is not clear what you mean by "fully streamed." What sort of request? Is it a `GET` or something with a body? Is the transfer encoding chunked or not?

**Reply by nflaig** [reply]
I mean a `POST` request with rather large request body (json). I know that for example when you use built-in fetch API you can access part of the request (like status code) before body is fully streamed and only `await res.json()` would require the body to be fully streamed (sorry if this is not the right term, maybe "fully received / transmitted / send"?)

In simple terms, I just want to figure out what is the time it took to transfer the payload over the wire by comparing a timestamp on my client against the timestamp on my server.

**Reply by jsumners** [reply]
`preValidation` would be the earliest hook with the body. See the notes on https://fastify.dev/docs/latest/Reference/Hooks/#preparsing

**Reply by nflaig** [reply]
But `preValidation` is after parsing stage so I would assume the request body was already parsed, i.e. in case of json it includes the time to do `JSON.parse`. What I am really looking for is just the network latency.

From the `preParsing` docs
>  you are using the preParsing hook, you can transform the request payload stream before it is parsed. It receives the request and reply objects as other hooks, and a stream with the current request payload.

This makes me think that when we call that hook the body is already fully streamed or can you transform a partially streamed body?



If you look at the example logs
```
07:38:50.098            debug: preParsing req-4a1 127.0.0.1 publishBlockV2
07:38:53.669            debug: preValidation req-4a1 127.0.0.1 publishBlockV2
```

What is causing the 3.5 seconds here? Is it parsing (`JSON.parse`) or does it still include time for streaming the request payload?

**Reply by jsumners** [reply]
Without a reproduction I will not venture a guess as to what is causing the time delta.
